tests/client.rs

```diff
@@ -1,88 +1,99 @@
-        .expect("send request");
-
-    assert_eq!(resp.status(), 200);
+    test_get("https://www.admin.ch/gov/de/start.html").await;
 }
 
 #[tokio::test]
 pub async fn test_get_many_headers() {
-    utils::setup_tracing();
-    let cache = utils::setup_cache();
-
-    let resp = Client::new(cache.path())
-        .await
-        .expect("create client")
-        .send(
-            Request::get("https://www.sunrise.ch/en/home")
-                .header("Host", "www.sunrise.ch")
-                .version(http::Version::HTTP_10)
-                .body(vec![])
-                .expect("create get request"),
-        )
-        .await
-        .expect("send request");
-
-    assert_eq!(resp.status(), 200);
+    test_get("https://www.sunrise.ch/en/home").await;
 }
 
 #[tokio::test]
 pub async fn test_post() {
+    test_client(
+        Request::post("https://httpbin.org/post")
+            .header("Host", "httpbin.org")
+            .version(http::Version::HTTP_10)
+            .body(Vec::from("key1=val1&key2=val2"))
+            .expect("Couldn't build request"),
+    )
+    .await;
+}
+
+// Creates a GET request from an URI and calls test_client on it.
+async fn test_get(uri: &str) {
+    let url = Url::parse(uri).unwrap();
+
+    test_client(
+        Request::get(uri)
+            .header("Host", url.host_str().unwrap())
+            .version(http::Version::HTTP_10)
+            .body(vec![])
+            .map_err(|e| anyhow::anyhow!("Error in request: {}", e))
+            .expect("Couldn't build request"),
+    )
+    .await;
+}
+
+// Calls the lightarti-rest code up to MAX_TRIES to get a correct answer.
+// Returns on the first correct answer, or throws a panic if all MAX_TRIES failed.
+// This is necessary due to the sometimes erratic behaviour of the tor-nodes.
+async fn test_client(req: Request<Vec<u8>>) {
     utils::setup_tracing();
-    let cache = utils::setup_cache();
+    let (header, body) = req.into_parts();
+    let host = header.uri.clone();
+
+    utils::test_n(MAX_TRIES, move || {
+        let request = clone_request(&header, &body);
+        Box::pin(async move {
+            let cache = utils::setup_cache();
+
+            let resp = Client::new(cache.path())
+                .await
+                .expect("create client")
+                .send(request)
+                .await;
 
-    let resp = Client::new(cache.path())
-        .await
-        .expect("create client")
-        .send(
-            Request::post("https://httpbin.org/post")
-                .header("Host", "httpbin.org")
-                .version(http::Version::HTTP_10)
-                .body(Vec::from("key1=val1&key2=val2"))
-                .expect("create post request"),
+            match resp {
+                Ok(resp) => (resp.status() == 200)
+                    .then_some(())
+                    .ok_or_else(|| anyhow::anyhow!("Wrong status")),
+                Err(e) => Err(anyhow::anyhow!("Couldn't get response: {}", e)),
+            }
```

Comment:
matching on `Result` is seen as bad practice if it can be indempotent

```suggestion
            resp.and_then(|r| {
                (r.status() == 200)
                    .then_some(())
                    .ok_or(anyhow::anyhow!("wrong status"))
            })
            .context("send request")
```

Reply:
here I don't want to use the `context`, as it has to return the error. But it's changed anyway with the inlining of `test_n`.